content/posts/2024-11-05-flang-new.md

```diff
@@ -0,0 +1,538 @@
+where we cannot assume that we started with a specific source language.
+
+# Flang Takes Driving Lessons
+
+**Note:** This section paraphrases material from
+["Flang Drivers"](https://github.com/llvm/llvm-project/blob/main/flang/docs/FlangDriver.md).
+If you want more detail please refer to that document.
+
+"Driver" in a compiler context means the part of the compiler that decides
+how to handle a set of options. For instance, when you use the option `-march=armv8a+memtag`,
+something in Flang knows that you want to compile for Armv8.0-a with the memory
+tagging extension enabled.
+
+`-march=` is an example of a "compiler driver" option. These options are what users
+give to the compiler. There is actually a second driver after this, confusingly
+called the "frontend" driver, despite being behind the scenes.
+
+In this case the "compiler driver" is `flang` and the "frontend driver" is
+`flang -fc1` (they are in the same binary).
+
+The main reason they are split into 2 is so that the "compiler driver" can have
+option names that do not change over time and the "frontend driver" can change
+its option names as suits the developers of the compiler.
+
+You can see the differences if you add `-###` to the compiler command:
+```
+$ ./bin/flang /tmp/test.f90 -march=armv8a+memtag -###
+ "<...>/flang" "-fc1"
+   "-triple" "aarch64-unknown-linux-gnu"
+   "-target-feature" "+v8a"
+   "-target-feature" "+mte"
+ "/usr/bin/ld" \
+   "-o" "a.out"
+   "-L/usr/lib/gcc/aarch64-linux-gnu/11"
+```
+(output reformatted and truncated for readability)
+
+The compiler driver has split the compilation into a job for the frontend
+(`flang -fc1`) and the linker (`ld`). `-march=` has been converted into many
+arguments to `flang -fc1`. This means that if compiler developers decided to
+change how `-march=` was converted, existing compile commands for `flang` would
+still work.
+
+Another compiler driver responsibility is to know where to find libraries on various
+systems. This differs between operating systems and even distributions of the
+same family of operating systems (for example Linux distributions).
+
+This created a problem when implementing Flang's compiler driver. This many
+details would take a long time to get right.
+
+Luckily for Flang, by this time it was in the LLVM Project, alongside Clang.
+Clang already knew how to handle this and had been tested on all sorts of
+systems over many years.
+
+> The intent is to mirror clang, for both the driver and CompilerInvocation, as
+> much as makes sense to do so. The aim is to avoid re-inventing the wheel and
+> to enable people who have worked with either the clang or flang entrypoints,
+> drivers, and frontends to easily understand the other.
+* [Peter Waller](https://discourse.llvm.org/t/rfc-adding-a-fortran-mode-to-the-clang-driver-for-flang/52307) (Arm)
+
+Flang became the the first project to use Clang's compiler driver library
+(`clangDriver`) to implement its own compiler driver.
```

Comment:
Welll ... IIRC, "Classic Flang" did the same, but that was out-of-tree. There might be other out-of-tree examples. But "LLVM Flang" remains the only in-tree example.

Reply:
I will note that, but given that Classic Flang is in the same lineage I think I can make almost the same statement still.

Reply:
I've said "first in tree".